### PR TITLE
improve performance of join by make it internal

### PR DESCRIPTION
**Pull request by wader**
Avoids exponential string append

Before:
BenchmarkJoinShort
BenchmarkJoinShort-4       26947             44399 ns/op
BenchmarkJoinLong
BenchmarkJoinLong-4          145           7295904 ns/op

After:
BenchmarkJoinShort
BenchmarkJoinShort-4   	   81789	     14456 ns/op
BenchmarkJoinLong
BenchmarkJoinLong-4    	   22002	     53791 ns/op

### CODE REVIEW

**Comment by wader** [comment]
Good question, yes should work, internal functions does not know about generators, will only get values. But i'm not sure how it get compiled to make it work in these cases, now i got curious. So i guess a call to a function that has any argument that results in no outputs will itself result in no output? seems resonable.

Tried your examples with the change:
```sh
$ go run cmd/gojq/main.go -nr '["a","b"] | join(",", "\n")'
a,b
a
b
$ go run cmd/gojq/main.go -nr '["a","b"] | join(empty)'
```

File: cli/test.yaml [hunk 1]
```diff
@@ -1777,7 +1777,7 @@
     - 'try join(",") catch .'
   input: '["1","2",[3,4,5]]'
   expected: |
-    "cannot add: string (\"1,2,\") and array ([3,4,5])"
+    "cannot join: array ([3,4,5])"
```
Comment (wader) [inline cli/test.yaml:1780]: I think it makes sense to change the error if we do it this way or?

**Comment by wader** [comment]
Maybe it's possible to have a more generic internal help function for improving performance of reduce + binop functions?  like keep map over array in jq and then pass result to a internal function with argument for how it should aggregate?

**Comment by itchyny** [comment]
You don't need to worry about stream as argument of internal functions because they are processed before called.
Don't break `{x:1,y:2} | join(",")`.

File: func.go [hunk 2]
```diff
@@ -1072,6 +1073,66 @@ func funcUniqueBy(v, x interface{}) interface{} {
...
+		ss := make([]string, len(v))
+		for i, e := range v {
+			if s, ok := funcJoinItem(e); ok {
+				ss[i] = s
+			} else {
+				return &unaryTypeError{name: "join", v: e}
+			}
+		}
+		return strings.Join(ss, sep)
+	case map[string]interface{}:
```
Comment (wader) [inline func.go:1112]: The object case got a bit messy. It does the same thing as `.[]`, each in sorted key order which is different from jq.

**Comment by itchyny** [comment]
This growing complexity is the reason I stopped making it internal. I don't like the duplicates with the executor and pathValue to be used by internal functions. So how about `def join($x): [.[]] | _join($x);` and making `_join` internal? Or `if type != "array" then [.[]] end | _join($x)` if we avoid cloning arrays.

File: func.go [hunk 3]
```diff
@@ -1072,6 +1073,47 @@ func funcUniqueBy(v, x interface{}) interface{} {
...
+		case int, float64, *big.Int:
+			ss[i] = jsonMarshal(e)
+		default:
+			return &unaryTypeError{name: "join", v: e}
+		}
+	}
+
+	if len(ss) == 1 {
+		return ss[0]
+	}
```
Comment (wader) [inline func.go:1107]: This is to allow `[1] | join(1)`, but would be nice to move this check earlier somehow

**Comment by wader** [comment]
Updated to use own string builder, long when from 57242 to 48389

**Comment by itchyny** [comment]
> By avoid cloning the array do you mean implement our own version of strings.Join so that we don't have to copy it?

No, use `strings.Join`. I just mean `[.[]] | _join($x)` clones the input array.

**Comment by wader** [comment]
Ok switching back to `strings.Join`

File: query_test.go [hunk 4]
```diff
@@ -241,6 +241,49 @@ func BenchmarkRun(b *testing.B) {
...
+			v, ok := iter.Next()
+			if !ok {
+				break
+			}
+			if _, ok := v.(error); ok {
+				b.Fatal(v)
+			}
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (wader) [reply query_test.go]: Ok, removing them

**Review by itchyny — APPROVED** [review]
Thank you.

**Comment by wader** [comment]
👍  appreciate your comments and suggestions

**Pull request merged**